util/src/main/java/io/camunda/zeebe/util/startup/StartupProcess.java

```diff
@@ -0,0 +1,230 @@
+   */
+  public StartupProcess(final Logger logger, final List<StartupStep<CONTEXT>> steps) {
+    this.steps = new ArrayDeque<>(Objects.requireNonNull(steps));
+    this.logger = Objects.requireNonNull(logger);
+  }
+
+  /**
+   * Executes the startup logic
+   *
+   * @param context the startup context at the start of this step
+   * @return future with startup context at the end of this step
+   */
+  public synchronized CompletableFuture<CONTEXT> startup(final CONTEXT context) {
+    logger.debug("Startup was called with context: " + context);
+    if (startupCalled) {
+      throw new IllegalStateException("startup(...) must only be called once");
+    }
+    startupCalled = true;
+
+    final var startupFuture = new CompletableFuture<CONTEXT>();
+    final var stepsToStart = new ArrayDeque<>(steps);
+
+    logger.info("Starting startup process");
+    proceedWithStartup(stepsToStart, context, startupFuture);
+
+    return startupFuture;
+  }
+
+  private synchronized void proceedWithStartup(
+      final Queue<StartupStep<CONTEXT>> stepsToStart,
+      final CONTEXT context,
+      final CompletableFuture<CONTEXT> startupFuture) {
+    if (shutdownFuture != null) {
+      logger.info("Aborting startup process because shutdown was called");
+      startupFuture.completeExceptionally(
+          new CancellationException("Aborting startup process because shutdown was called"));
+    } else if (stepsToStart.isEmpty()) {
+      startupFuture.complete(context);
+      logger.info("Finished startup process");
+    } else {
+      final var stepToStart = stepsToStart.poll();
+      startedSteps.push(stepToStart);
+
+      logCurrentStep("Startup", stepToStart);
+      stepToStart
+          .startup(context)
+          .whenComplete(
+              (contextReturnedByStep, error) -> {
+                if (error != null) {
+                  logger.warn(
+                      "Aborting startup process due to exception during stage "
+                          + stepToStart.getName(),
+                      error);
+                  startupFuture.completeExceptionally(error);
+                } else {
+                  proceedWithStartup(stepsToStart, contextReturnedByStep, startupFuture);
+                }
+              });
+    }
+  }
+
+  /**
+   * Executes the shutdown logic
+   *
+   * @param context the shutdown context at the start of this step
+   * @return future with the shutdown context at the end of this step.
+   */
+  public synchronized CompletableFuture<CONTEXT> shutdown(final CONTEXT context) {
+    logger.debug("Shutdown was called with context: " + context);
+    if (shutdownFuture == null) {
+      logger.info("Starting shutdown process");
+      shutdownFuture = new CompletableFuture<>();
+
+      proceedWithShutdown(context, shutdownFuture, new ArrayList<>());
+    } else {
+      logger.info("Shutdown already in progress");
+    }
+    return shutdownFuture;
+  }
+
+  private void proceedWithShutdown(
```

Comment:
:x: I think we may run into similar issues as with `proceedWithStartup`. Since `startedSteps` is not thread safe, it's possible the caller thread does not see the latest value (as we pop here, and then recurse potentially in another thread causing us to check there if `startedSteps` is empty). The options are either to synchronize `proceedWithShutdown`, pass the whole context of the method as parameter (e.g. startedSteps would also become a parameter so we always have the latest value), or make it a thread safe object. 

I think this is very unlikely though, so I couldn't really reproduce this, but it sounds plausible to me from experience. wdyt?